test-suite/js/src/test/scala/org/scalajs/testsuite/jsinterop/NestedJSClassTest.scala

```diff
@@ -0,0 +1,596 @@
+    assertTrue(js.special.instanceof(inner3, localJSClass2))
+    assertFalse(js.special.instanceof(inner3, localJSClass1))
+    assertFalse(inner3.isInstanceOf[container1.InnerJSClass])
+  }
+
+  @Test def innerJSObject_basics(): Unit = {
+    val container1 = new ScalaClassContainerWithObject("hello")
+    val inner1 = container1.InnerJSObject
+    assertSame(inner1, container1.InnerJSObject)
+    assertEquals("object", js.typeOf(inner1))
+
+    assertEquals("hellozzz", inner1.zzz)
+    assertEquals("hellozzzfoo", inner1.foo("foo"))
+    assertTrue(inner1.isInstanceOf[container1.InnerJSObject.type])
+
+    val container2 = new ScalaClassContainerWithObject("hi")
+    val inner2 = container2.InnerJSObject
+    assertNotSame(inner1, inner2)
+    assertNotSame(inner1.asInstanceOf[js.Dynamic].constructor,
+        inner2.asInstanceOf[js.Dynamic].constructor)
+    assertEquals("hizzz", inner2.zzz)
+    assertEquals("hizzzfoo", inner2.foo("foo"))
+
+    assertFalse(inner2.isInstanceOf[container1.InnerJSObject.type])
+  }
+
+  @Test def localJSObject_basics(): Unit = {
+    val container1 = new ScalaClassContainerWithObject("hello")
+    val inner1 = container1.makeLocalJSObject("world1")
+
+    assertEquals("helloworld1", inner1.zzz)
+    assertEquals("helloworld1foo", inner1.foo("foo"))
+
+    val inner2 = container1.makeLocalJSObject("world2")
+    assertEquals("helloworld2", inner2.zzz)
+    assertEquals("helloworld2foo", inner2.foo("foo"))
+
+    assertNotSame(inner1, inner2)
+    assertNotSame(inner1.asInstanceOf[js.Dynamic].constructor,
+        inner2.asInstanceOf[js.Dynamic].constructor)
+  }
+
+  @Test def innerJSClassExtendsInnerJSClass(): Unit = {
+    val parentsContainer = new ScalaClassContainer("hello")
+    val container1 =
+      new ScalaClassContainerWithSubclasses("hi", parentsContainer)
+    val innerJSClass = parentsContainer.getInnerJSClass
+    val innerJSSubclass = container1.getInnerJSSubclass
+
+    val inner1 = new container1.InnerJSSubclass("world1")
+    assertEquals("helloworld1", inner1.zzz)
+    assertEquals("helloworld1foo", inner1.foo("foo"))
+    assertEquals("hiworld1helloworld1", inner1.foobar())
+
+    assertTrue(inner1.isInstanceOf[container1.InnerJSSubclass])
+    assertTrue(js.special.instanceof(inner1, innerJSSubclass))
+    assertTrue(inner1.isInstanceOf[parentsContainer.InnerJSClass])
+    assertTrue(js.special.instanceof(inner1, innerJSClass))
+
+    val container2 =
+      new ScalaClassContainerWithSubclasses("salut", parentsContainer)
+    val innerJSSubclass2 = container2.getInnerJSSubclass
+
+    val inner2 = js.Dynamic.newInstance(innerJSSubclass2)("world2")
+    assertEquals("helloworld2", inner2.zzz)
+    assertEquals("helloworld2foo", inner2.foo("foo"))
+    assertEquals("salutworld2helloworld2", inner2.foobar())
+
+    assertTrue((inner2: Any).isInstanceOf[container2.InnerJSSubclass])
+    assertFalse((inner2: Any).isInstanceOf[container1.InnerJSSubclass])
```

Comment:
Why does this need an ascription to `Any`?

Reply:
To avoid the compiler warning "inner2 with type `js.Dynamic` can never be an instance of `container2.InnerJSSubclass`"